base/abstractarray.jl

```diff
@@ -66,15 +65,16 @@ is `indices(A, 1)`.
 Calling this function is the "safe" way to write algorithms that
 exploit linear indexing.
 """
-linearindices(A)                 = (@_inline_meta; 1:length(A))
+linearindices(A)                 = (@_inline_meta; OneTo(unsafe_length(A)))
 linearindices(A::AbstractVector) = (@_inline_meta; indices1(A))
 eltype{T}(::Type{AbstractArray{T}}) = T
 eltype{T,N}(::Type{AbstractArray{T,N}}) = T
 elsize{T}(::AbstractArray{T}) = sizeof(T)
 ndims{T,N}(::AbstractArray{T,N}) = N
 ndims{T,N}(::Type{AbstractArray{T,N}}) = N
 ndims{T<:AbstractArray}(::Type{T}) = ndims(supertype(T))
-length(t::AbstractArray) = prod(size(t))::Int
+length(t::AbstractArray) = prod(size(t))
+unsafe_length(A::AbstractArray) = prod(map(unsafe_length, indices(A)))  # internal method
```

Comment:
different names for different concepts is usually a good policy? though we aren't super consistent yet about what `unsafe_` means as a general guideline